Case API v.2: Upsert on `external_id`

kaapstorm opened the pull request:
## Technical Summary

Some context: [SUPPORT-25970](https://dimagi.atlassian.net/browse/SUPPORT-25970)

(SaaS/CommCare ticket pending)

Adds upsert functionality on `external_id` for Case API v.2 [Bulk Create/Update Cases](https://github.com/dimagi/commcare-hq/blob/94d4e8ff0dfdc4c13a20101ef0e6284b0390cc37/docs/api/cases-v2.rst#bulk-createupdate-cases).

See the [update to documentation](https://github.com/dimagi/commcare-hq/commit/c7e8be209cfad0818b5cdc254cf25be100363c16) for details and warnings.

Curious to hear your thoughts on this @esoergel 

## Safety Assurance

### Safety story

* This change is limited to the Case API v.2
* Change in behavior is only encountered if users do not follow existing documentation.

### Automated test coverage

Yes

### QA Plan

QA planned. Ticket pending.


### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SUPPORT-25970]: https://dimagi.atlassian.net/browse/SUPPORT-25970?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kaapstorm committed:
Add upsert on external_id to _get_bulk_updates()

kaapstorm committed:
Tests

kaapstorm committed:
Documentation

kaapstorm committed:
Move upsert logic into `_get_individual_update()`

kaapstorm committed:
Update tests

esoergel commented on corehq/apps/hqcase/api/updates.py:
```diff
@@ -199,10 +226,9 @@ def _get_bulk_updates(domain, all_data, user):
+            if 'create' not in data:
                 raise UserError("A 'create' flag is required for each update.")
-            updates.append(_get_individual_update(domain, data, user, is_creation))
+            updates.append(_get_individual_update(domain, data, user, is_creation=True))
```
I think I prefer either
```suggestion
            updates.append(_get_individual_update(domain, data, user, is_creation=data.pop('create')))
```
or making the other caller of `_get_individual_update` insert `create` into `data`, so there's only one way to pass it in.

kaapstorm committed:
Pythonic asserts

kaapstorm committed:
Add tests for `get_bulk()`

kaapstorm committed:
Add GET endpoint for external_id

kaapstorm committed:
Implement PUT for external_id

kaapstorm committed:
Move UPSERT logic back to `_get_bulk_updates()`

kaapstorm committed:
Add and update tests

esoergel commented on corehq/apps/hqcase/tests/test_case_api_bulk_get.py:
```diff
@@ -87,6 +87,41 @@ def test_get_single_case(self):
         assert res.status_code == 200
         assert res.json()['case_id'] == self.case_ids[0]
 
+    def test_get_single_case_by_external_id(self):
```
Nit: this file is called `test_case_api_bulk_get`, but this functionality is not part of that

kaapstorm replied:
61b7de6e904

esoergel commented on corehq/apps/hqcase/tests/test_case_api_bulk_get.py:
```diff
@@ -87,6 +87,41 @@ def test_get_single_case(self):
+        base_url = reverse('case_api', args=(self.domain,))
+        if not base_url.endswith('/'):
+            base_url += '/'
+        url = f"{base_url}ext/{external_id}/"
```
Why not just name the external ID urlpattern and reverse that?  Looks like you could even give it the same name as long as you reverse it with `kwargs` instead of args:
```python
>>> reverse('case_api', kwargs={'domain': 'esoergel', 'external_id': 'abc123'})
'/a/esoergel/api/v0.6/case/ext/abc123/'
>>> reverse('case_api', kwargs={'domain': 'esoergel', 'case_id': 'abc123'})
'/a/esoergel/api/case/v2/abc123'
```
Though that feels more ambiguous - a separate name would be clearer, I think

esoergel commented on corehq/apps/hqcase/views.py:
```diff
@@ -193,10 +193,42 @@ def _handle_list_view(request):
+    if case.get('error') == 'not found':
+        is_creation = True
+    else:
+        is_creation = False
```
I didn't think anything of it with the `_handle_ext_get` function, but here I'm wondering if this wouldn't be better served with something bespoke, either extracted from `get_bulk`'s internals or paralleling it.  There's a lot that goes on there that doesn't apply here, and it makes this code a bit weird - looks like it can raise a `UserError`, but at present that's only for page size issues.  Reading the error message from an API call seems like a convoluted way to go about seeing if a case is present, and what if there's another error returned?

Instead, you could just do a simple get by external ID (failing if there are multiple matches) and run it past `user_can_access_case` for location safety.  That could be a util used here and in `_handle_ext_get`.

esoergel commented on corehq/apps/hqcase/views.py:
```diff
@@ -193,10 +193,42 @@ def _handle_list_view(request):
+                          f'external_id = "{external_id}". "case_id" is '
+                          'read-only and cannot be modified.'},
+                status=400,
+            )
```
Good idea to call this out specifically

esoergel commented on corehq/apps/hqcase/tests/test_api_updates.py:
```diff
@@ -0,0 +1,233 @@
+
+
+@sharded
+class TestIndividualUpdate(TestCase):
```
For what this test covers, I don't think it needs to be a DB test - it doesn't actually submit any forms or create cases, so the `tearDown` is superfluous, and it looks like `_get_individual_update` doesn't actually even use the `domain` arg (must be leftover from some old code shuffling).  All it needs `user` for is for the user ID, so that could easily be mocked, or switch the arg to `user_id`

kaapstorm replied:
b717b1c9d6b

kaapstorm committed:
isort

kaapstorm committed:
JsonCaseUpsert

kaapstorm committed:
Add tests

kaapstorm committed:
Pythonic asserts

esoergel reviewed (commented):
I have a bunch of open comments from my last review, but I think they're all pretty nitpicky and small.  I like the big picture approach overall

kaapstorm committed:
Raise error if is_new_case not initialized

kaapstorm commented on corehq/apps/hqcase/api/updates.py:
```diff
@@ -158,6 +158,37 @@ def get_case_id(self, case_db):
+    case_type = jsonobject.StringProperty(required=True)
+    owner_id = jsonobject.StringProperty(required=True)
+
+    _is_case_creation = None  # Determined when get_case_id() is called
```
Good call! 03da5db81fe

kaapstorm committed:
Check required properties late

kaapstorm committed:
Use path converter for `external_id`

kaapstorm committed:
Show behavior with repeated results

kaapstorm commented on corehq/apps/hqcase/tests/test_case_api_bulk_get.py:
```diff
@@ -218,3 +220,113 @@ def _check_matching_missing(self, result, matching=None, missing=None):
+    def test_get_bulk_duplicate_external_id(self):
+        result = get_bulk(self.domain, self.web_user, external_ids=['test2'])
+
+        # When multiple cases have the same external_id, only one is
```
Yeah. [b5e63234eda](https://github.com/dimagi/commcare-hq/commit/b5e63234eda8582342cb87bca9f56dd7ebdb844d)

To fix `BulkFetchResults.merge()` we could use a dict keyed on case_id instead of a list for `BulkFetchResults.cases`, but we should probably drop `BulkFetchResults.missing_records`. It is based on IDs of different kinds. If a case is not found by case_id, but it is found by external_id, is it missing? That would change the API return value format. :face_with_diagonal_mouth: Feels out of scope for this change.

kaapstorm committed:
Simplify URL patterns

Django's
[APPEND_SLASH](https://docs.djangoproject.com/en/5.2/ref/settings/#append-slash)
setting will append a trailing slash, so we don't have to accommodate
missing trailing slashes.

kaapstorm committed:
Simplify function params

kaapstorm committed:
Drop redundant checks

kaapstorm commented on corehq/apps/hqcase/api/updates.py:
```diff
@@ -191,9 +222,19 @@ def _get_individual_update(domain, data, user, is_creation):
+    if 'case_id' in data:
+        raise UserError("UPSERT does not allow case_id to be specified")
+    if not data.get('external_id'):
+        raise UserError("UPSERT requires external_id to be specified")
```
Yes. Thank you. 3fb336082ee

kaapstorm committed:
Simplify tests

kaapstorm committed:
Consolidate tests

kaapstorm committed:
Pythonic asserts

kaapstorm committed:
Separate out single-case GET operations

kaapstorm commented on corehq/apps/hqcase/tests/test_api_updates.py:
```diff
@@ -0,0 +1,233 @@
+
+    def tearDown(self):
+        FormProcessorTestUtils.delete_all_cases(self.domain)
+        FormProcessorTestUtils.delete_all_xforms(self.domain)
```
b717b1c9d6b

kaapstorm committed:
Merge branch 'master' into nh/case_api_upsert

kaapstorm commented:
@esoergel I think my recent changes cover everything. Please let me know if there is anything I've missed, or if there is anything else you think should change.

kaapstorm committed:
Lint

esoergel commented on corehq/apps/api/urls.py:
```diff
@@ -143,6 +143,8 @@ def versioned_apis(api_list):
     # match v0.6/case/ AND v0.6/case/e0ad6c2e-514c-4c2b-85a7-da35bbeb1ff1/ trailing slash optional
```
Thanks!  I think I didn't remember at the time that you can use the same name for multiple patterns (for reversing, say)

esoergel commented on corehq/apps/hqcase/tests/test_case_api_get.py:
```diff
@@ -0,0 +1,124 @@
+@disable_quickcache
+@privilege_enabled(privileges.API_ACCESS)
+@flag_enabled('API_THROTTLE_WHITELIST')
+class TestCaseAPIGet(TestCase):
```
oof, this does add a lot of extra test code, huh?

kaapstorm replied:
Cleaning up tests, and cleaning up code to make it more testable as I go, is an aspiration. :upside_down_face: :sparkles:

esoergel reviewed (commented):
Just one blocker here with the trailing slash thing, but otherwise looks great!

kaapstorm committed:
Support optional trailing slash

kaapstorm commented on corehq/apps/hqcase/tests/test_case_update_api.py:
```diff
@@ -692,5 +692,5 @@ def test_urls_without_trailing_slash(self):
             )
-            # These requests should return a 400 because of the bad body, not a 301 redirect
-            self.assertEqual(res.status_code, 400)
+            # Django's APPEND_SLASH setting redirects URLs without trailing slashes
```
Ooo! Good to know! 6b715d16a39

kaapstorm committed:
Check PUT payload is a dict

kaapstorm committed:
Add tests
